src/Arcus.WebApi.Security/Authentication/CertificateAuthenticationValidator.cs

```diff
@@ -0,0 +1,126 @@
+        {
+            Guard.NotNull(certificateAuthenticationConfig, nameof(certificateAuthenticationConfig), "Certificate authentication configuration cannot be 'null'");
+
+            _certificateAuthenticationConfig = certificateAuthenticationConfig;
+        }
+
+        /// <summary>
+        /// Validate the specified <paramref name="clientCertificate"/> based on the configured certificate requirements
+        /// using the <paramref name="services"/> to retrieve registered instances that will provide the expected certificate values.
+        /// </summary>
+        /// <param name="clientCertificate">The client certificate to validate.</param>
+        /// <param name="services">The collection of registered services, (ex. from the request pipeline).</param>
+        /// <returns>
+        ///     <c>true</c> when the specified <paramref name="clientCertificate"/> is valid, <c>false</c> otherwise.
+        /// </returns>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="clientCertificate"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="services"/> is <c>null</c>.</exception>
+        internal async Task<bool> ValidateCertificate(X509Certificate2 clientCertificate, IServiceProvider services)
+        {
+            Guard.NotNull(clientCertificate, nameof(clientCertificate), "Certificate authentication validation requires a client certificate");
+            Guard.NotNull(services, nameof(services), "Certificate authentication validation requires a service object to retrieve registered services");
+
+            ILogger logger = GetLoggerOrDefault(services);
+
+            IDictionary<X509ValidationRequirement, ExpectedCertificateValue> expectedValuesByRequirement =
+                await _certificateAuthenticationConfig.GetAllExpectedCertificateValues(services, logger);
+
+            return expectedValuesByRequirement.All(keyValue =>
+            {
+                switch (keyValue.Key)
+                {
+                    case X509ValidationRequirement.SubjectName:
+                        return IsCertificateSubjectNameAllowed(clientCertificate, keyValue.Value, logger);
+                    case X509ValidationRequirement.IssuerName:
+                        return IsCertificateIssuerNameAllowed(clientCertificate, keyValue.Value, logger);
+                    case X509ValidationRequirement.Thumbprint:
+                        return IsCertificateThumbprintAllowed(clientCertificate, keyValue.Value, logger);
+                    default:
+                        throw new ArgumentOutOfRangeException(nameof(keyValue.Key), keyValue.Key, "Unknown validation type specified");
+                }
+            });
+        }
+
+        private static ILogger GetLoggerOrDefault(IServiceProvider services)
+        {
+            return services.GetService<ILoggerFactory>()
```

Comment:
Would be more readable to capture output of Create Logger and do the short hand return if not null on a dedicated line.

This whole method is one statement which is an anti-pattern imo

Reply:
Not sure if I share that opinion of it being an anti-pattern since it allows devs to more easily add null-checks. Which I would consider a plus.

You can argue of course about readability and such; certainly if we want to express our Arcus code-base in the most basic possible way to not lose beginner devs in syntax.

Reply:
I was more referring to this as it's more readable, if you all do it on one line it's not really reable
```csharp
ILogger logger = services.GetService<ILoggerFactory>()
                         ?.CreateLogger<CertificateAuthenticationFilter>();

return logger ?? NullLogger.Instance;
```

Reply:
But current approach is ok as well, no worries